### PR TITLE
Adding shortread deduplication feature with fastp

### PR DESCRIPTION
**Pull request by maxibor**
This PR adds deduplication of reads with fastp

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Check for unexpected warnings in debug mode (`nextflow run . -profile debug,test,docker --outdir <OUTDIR>`).
- [ ] `CHANGELOG.md` is updated.

### CODE REVIEW

**Comment by maxibor** [comment]
An easy PR for you @jfy133 to offset [#548](https://github.com/nf-core/mag/pull/548) 😉

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### `Added`
 
+- [#439](https://github.com/nf-core/taxprofiler/pull/439) Read deduplication with fastp
```
Comment (jfy133) [inline CHANGELOG.md:10]:
```suggestion
- [#439](https://github.com/nf-core/taxprofiler/pull/439) Read deduplication with fastp (added by @maxibor)
```

File: nextflow_schema.json [hunk 2]
```diff
@@ -147,6 +147,12 @@
                     "description": "Specify the minimum length of reads to be retained",
                     "help_text": "Specifying a mimum read length filtering can speed up profiling by reducing the number of short unspecific reads that need to be match/aligned to the database.\n\n> Modifies tool parameter(s):\n> - removed from reads `--length_required`\n> - AdapterRemoval: `--minlength`"
                 },
+                "shortread_qc_dedup": {
+                    "type": "boolean",
+                    "fa_icon": "fas fa-toggle-on",
+                    "description": "Perform deduplication of the input reads. Only when shortread QC is done with fastp",
```
Comment (jfy133) [inline nextflow_schema.json:153]:
```suggestion
                    "description": "Perform deduplication of the input reads (fastp only)",
```

File: nextflow_schema.json [hunk 3]
```diff
@@ -147,6 +147,12 @@
...
+                    "help_text": "This enable the deduplication to drop the duplicated reads/pairs with fastp"
```
Comment (jfy133) [inline nextflow_schema.json:154]:
```suggestion
                    "help_text": "This enables the deduplication of processed reads during fastp adapter removal and/or merging. It removes identical reads that are likely artefacts from laboratory protocols (e.g. amplification), and provide no additional sequence information to the library.\n\nRemoving duplicates can increase runtime and increase accuracy of abundance calculations.\n\n> Modifies tool parameter(S):\n> fastp: ` --dedup`\n"
```

**Review by jfy133 — COMMENTED** [review]
Indeed this is much easier :grimacing: 

Doc improvements but otherwise LGTM. @sofstam @LilyAnderssonLee should we add here as 1.5 or to bouncy basenji? It's not a bug fix exactly... but it's not a fully fledged new functionality addition?

We do have potentially teh centrifuge fix coming up (if I can fix it...) so there likely will be a 1.1.5

**Comment by sofstam** [comment]
I would say it is 1.1.5.

**Comment by sofstam** [comment]
Getting back with a review later today :)

**Review by sofstam — APPROVED** [review]

**Comment by Midnighter** [comment]
I'm somewhat concerned with this feature. If I remember correctly, FASTQC and fastp use the first 50 and 75 bp, respectively, to judge read duplication. Using longer sequences would drive up memory requirements and take longer. So the first question is, are we truly only removing identical reads with this?

My second question comes from my inexperience with sequencing: If you have a dominant species in your metagenomic sample, how unlikely is it to have an identical read?

**Comment by jfy133** [comment]
> sing longer sequences would drive up memory requirements and take longer. So the first question is, are we truly only removing identical reads with this?

Does it really? the README at least seems to implies it's some condensed hash of the whole read: https://github.com/OpenGene/fastp#duplication-rate-evaluation. That said, it's opt-in so it's still up to the user to decide if it's suitable

> My second question comes from my inexperience with sequencing: If you have a dominant species in your metagenomic sample, how unlikely is it to have an identical read?

An absolutely exact duplicate is quite unlikely, as

1. fragmentation protocols should be random (with a slight preference breakages around GCs IIRC), so that in combination with (relatively) longer reads it's unlikely due to sequence diversity. 

2. Exact duplicates are _much_ more likely from lab-based amplicons as they use the same priming sequence, and given the number of amplification cycles also very likely to have copies from artifical duplicate rather than naturally occuring.

**Comment by Midnighter** [comment]
Thank you for your response, sounds good to me then. 👍🏼

File: nextflow_schema.json [hunk 4]
```diff
@@ -147,6 +147,12 @@
                     "description": "Specify the minimum length of reads to be retained",
                     "help_text": "Specifying a mimum read length filtering can speed up profiling by reducing the number of short unspecific reads that need to be match/aligned to the database.\n\n> Modifies tool parameter(s):\n> - removed from reads `--length_required`\n> - AdapterRemoval: `--minlength`"
                 },
+                "shortread_qc_dedup": {
+                    "type": "boolean",
+                    "fa_icon": "fas fa-toggle-on",
+                    "description": "Perform deduplication of the input reads (fastp only)",
+                    "help_text": "This enables the deduplication of processed reads during fastp adapter removal and/or merging. It removes identical reads that are likely artefacts from laboratory protocols (e.g. amplification), and provide no additional sequence information to the library.\n\nRemoving duplicates can increase runtime and increase accuracy of abundance calculations.\n\n> Modifies tool parameter(S):\n> fastp: ` --dedup`\n"
```
Comment (jfy133) [inline nextflow_schema.json:154]:
```suggestion
                    "help_text": "This enables the deduplication of processed reads during fastp adapter removal and/or merging. It removes identical reads that are likely artefacts from laboratory protocols (e.g. amplification), and provide no additional sequence information to the library.\n\nRemoving duplicates can increase runtime and increase accuracy of abundance calculations.\n\n> Modifies tool parameter(s):\n> fastp: ` --dedup`\n"
```

**Pull request merged**